# Практическая работа

# Задача

Один из способов повысить эффективность взаимодействия банка с клиентами — отправлять предложение о новой услуге не всем клиентам, а только некоторым, которые выбираются по принципу наибольшей склонности к отклику на это предложение.

Задача заключается в том, чтобы предложить алгоритм, который будет выдавать склонность клиента к положительному или отрицательному отклику на предложение банка. Предполагается, что, получив такие оценки для некоторого множества клиентов, банк обратится с предложением только к тем, от кого ожидается положительный отклик.


Для решения этой задачи загрузите файлы из базы в Postgres.
Эта БД хранит информацию о клиентах банка и их персональные данные, такие как пол, количество детей и другие.

Описание таблиц с данными представлено ниже.


**D_work**

Описание статусов относительно работы:
- ID — идентификатор социального статуса клиента относительно работы;
- COMMENT — расшифровка статуса.


**D_pens**

Описание статусов относительно пенсии:
- ID — идентификатор социального статуса;
- COMMENT — расшифровка статуса.


**D_clients**

Описание данных клиентов:
- ID — идентификатор записи;
- AGE	— возраст клиента;
- GENDER — пол клиента (1 — мужчина, 0 — женщина);
- EDUCATION — образование;
- MARITAL_STATUS — семейное положение;
- CHILD_TOTAL	— количество детей клиента;
- DEPENDANTS — количество иждивенцев клиента;
- SOCSTATUS_WORK_FL	— социальный статус клиента относительно работы (1 — работает, 0 — не работает);
- SOCSTATUS_PENS_FL	— социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
- REG_ADDRESS_PROVINCE — область регистрации клиента;
- FACT_ADDRESS_PROVINCE — область фактического пребывания клиента;
- POSTAL_ADDRESS_PROVINCE — почтовый адрес области;
- FL_PRESENCE_FL — наличие в собственности квартиры (1 — есть, 0 — нет);
- OWN_AUTO — количество автомобилей в собственности.


**D_agreement**

Таблица с зафиксированными откликами клиентов на предложения банка:
- AGREEMENT_RK — уникальный идентификатор объекта в выборке;
- ID_CLIENT — идентификатор клиента;
- TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было).
    
    
**D_job**

Описание информации о работе клиентов:
- GEN_INDUSTRY — отрасль работы клиента;
- GEN_TITLE — должность;
- JOB_DIR — направление деятельности внутри компании;
- WORK_TIME — время работы на текущем месте (в месяцах);
- ID_CLIENT — идентификатор клиента.


**D_salary**

Описание информации о заработной плате клиентов:
- ID_CLIENT — идентификатор клиента;
- FAMILY_INCOME — семейный доход (несколько категорий);
- PERSONAL_INCOME — личный доход клиента (в рублях).


**D_last_credit**

Информация о последнем займе клиента:
- ID_CLIENT — идентификатор клиента;
- CREDIT — сумма последнего кредита клиента (в рублях);
- TERM — срок кредита;
- FST_PAYMENT — первоначальный взнос (в рублях).


**D_loan**

Информация о кредитной истории клиента:
- ID_CLIENT — идентификатор клиента;
- ID_LOAN — идентификатор кредита.

**D_close_loan**

Информация о статусах кредита (ссуд):
- ID_LOAN — идентификатор кредита;
- CLOSED_FL — текущий статус кредита (1 — закрыт, 0 — не закрыт).

Ниже представлен минимальный список колонок, которые должны находиться в итоговом датасете после склейки и агрегации данных. По своему усмотрению вы можете добавить дополнительные к этим колонки.

    - AGREEMENT_RK — уникальный идентификатор объекта в выборке;
    - TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было);
    - AGE — возраст клиента;
    - SOCSTATUS_WORK_FL — социальный статус клиента относительно работы (1 — работает, 0 — не работает);
    - SOCSTATUS_PENS_FL — социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
    - GENDER — пол клиента (1 — мужчина, 0 — женщина);
    - CHILD_TOTAL — количество детей клиента;
    - DEPENDANTS — количество иждивенцев клиента;
    - PERSONAL_INCOME — личный доход клиента (в рублях);
    - LOAN_NUM_TOTAL — количество ссуд клиента;
    - LOAN_NUM_CLOSED — количество погашенных ссуд клиента.


Будьте внимательны при сборке датасета: это реальные банковские данные, в которых могут наблюдаться дубли, некорректно заполненные значения или значения, противоречащие друг другу. Для получения качественной модели необходимо предварительно очистить датасет от такой информации.

## Задание 1

В предыдущем задании вы собрали всю информацию о клиентах в одну таблицу, где одна строчка соответствует полной информации об одном клиенте.

Загрузите эту таблицу.

In [1]:
# your code here
import pandas as pd

In [2]:
df = pd.read_csv('datasets/bank_clients.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15594 entries, 0 to 15593
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   AGREEMENT_RK       15594 non-null  int64  
 1   AGE                15594 non-null  int64  
 2   SOCSTATUS_WORK_FL  15594 non-null  int64  
 3   SOCSTATUS_PENS_FL  15594 non-null  int64  
 4   GENDER             15594 non-null  int64  
 5   CHILD_TOTAL        15594 non-null  int64  
 6   DEPENDANTS         15594 non-null  int64  
 7   PERSONAL_INCOME    15594 non-null  float64
 8   LOAN_NUM_TOTAL     15594 non-null  int64  
 9   LOAN_NUM_CLOSED    15594 non-null  int64  
 10  TARGET             15594 non-null  int64  
dtypes: float64(1), int64(10)
memory usage: 1.3 MB


In [3]:
df.head()

AGREEMENT_RK  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  GENDER  \
0      60099204   42                  1                  0       1   
1      62244665   28                  1                  0       1   
2      61050759   64                  1                  1       0   
3      62079659   54                  1                  0       1   
4      66583553   26                  1                  0       0   

   CHILD_TOTAL  DEPENDANTS  PERSONAL_INCOME  LOAN_NUM_TOTAL  LOAN_NUM_CLOSED  \
0            1           0          25000.0               1                1   
1            1           1          10000.0               1                1   
2            2           0          30000.0               1                1   
3            0           0          25000.0               1                1   
4            1           1          15000.0               2                2   

   TARGET  
0       1  
1       0  
2       0  
3       0  
4       0

Разбейте данные на тренировочную и тестовую часть в пропорции 80% к 20%, зафиксируйте `random_state = 42`.

In [4]:
# your code here
X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X , y, test_size=0.2, random_state=42)

In [6]:
#pip install "numpy>=1.16.5,<1.25.0"

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.3/17.3 MB 10.6 MB/s eta 0:00:00m eta 0:00:010:01:01
  Attempting uninstall: numpy
    Found existing installation: numpy 1.26.4
    Uninstalling numpy-1.26.4:
      Successfully uninstalled numpy-1.26.4
Note: you may need to restart the kernel to use updated packages.


На тренировочных данных обучите линейную модель классификации для предсказания целевой переменной (столбец `TARGET`).

Сделайте прогноз вероятности отклика на рекламную кампанию для тестовых данных.

In [6]:
# your code here
from sklearn.linear_model import LogisticRegression

model =LogisticRegression()
model.fit(X_train, y_train)
pred_test = model.predict(X_test) # предсказание класса

/home/olegerius/.local/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Переведите вероятности в классы по стандартному порогу (0.5) и на тестовом наборе данных вычислите метрики:

* accuracy
* precision
* recall
* f1-score

In [7]:
# your code here
pred_test[:25]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0])

In [8]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

accuracy_score(y_test, pred_test)

0.8781660788714332

In [9]:
y.value_counts(normalize=True)

TARGET
0    0.883801
1    0.116199
Name: proportion, dtype: float64

имеется дисбаланс классов

In [10]:
confusion_matrix(y_test, pred_test)

array([[2739,    4],
       [ 376,    0]])

### *Для 2741 клиентов, которые не откликнулись модель ошибочно предсказала двоих. 376 клиентов которые сотрудничали и откликнулись, модель записала в неоткликнувшиеся*

In [11]:
prob = model.predict_proba(X_test)
prob[:7]

array([[0.88002147, 0.11997853],
       [0.92029709, 0.07970291],
       [0.9255957 , 0.0744043 ],
       [0.88182655, 0.11817345],
       [0.84472668, 0.15527332],
       [0.81028543, 0.18971457],
       [0.894186  , 0.105814  ]])

In [12]:
prob_churn = prob[:,1]
classes = prob_churn > 0.5

In [13]:
print(f"Accuracy: {accuracy_score(y_test, classes)}")
print(f"Precision: {precision_score(y_test, classes)}")
print(f"Recall: {recall_score(y_test, classes)}")
print(f"f1: {f1_score(y_test, classes)}")

Accuracy: 0.8781660788714332
Precision: 0.0
Recall: 0.0
f1: 0.0


In [14]:
confusion_matrix(y_test, classes)

array([[2739,    4],
       [ 376,    0]])

Целевая метрика для задачи - полнота, так как нам нужно найти максимум клиентов, кто может откликнуться на рекламу.

Но при этом точность не должна просесть, поэтому за ней тоже следим.

Разбейте тренировочные данные на `train` и `val` части в пропорции 3 к 1.

В цикле:

* переберите пороги от 0 до 1 с шагом 0.01
* вычислите для каждого порога значение метрик precision и recall
* подберите такой порог, при котором recall не меньше 0.66, а точность максимальна.

In [15]:
# your code here
df = pd.read_csv('datasets/bank_clients.csv')

X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']

X_train, X_val, y_train, y_val = train_test_split(X , y, test_size=0.25, random_state=42)





In [16]:
max_precision = -1
border = 0
i = 0

model =LogisticRegression()
model.fit(X_train, y_train)
pred_test = model.predict(X_val)

prob = model.predict_proba(X_val)
prob_churn = prob[:,1]

while i <=1:
  classes = prob_churn > i
  if (recall_score(y_val, classes) >= 0.66) and (precision_score(y_val, classes) > max_precision):
    border = i
    max_precision = precision_score(y_val, classes)
  i += 0.01




/home/olegerius/.local/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [17]:
border, max_precision

(0.09999999999999999, 0.14450354609929078)

Для выбранного порога посчитайте все метрики на тестовых данных. Сильно ли они отличаются от метрик на валидации?

In [18]:
# your code here
classes = prob_churn > border

print(f"Accuracy: {accuracy_score(y_val, classes)}")
print(f"Precision: {precision_score(y_val, classes)}")
print(f"Recall: {recall_score(y_val, classes)}")
print(f"f1: {f1_score(y_val, classes)}")


Accuracy: 0.47063349576814567
Precision: 0.14450354609929078
Recall: 0.7086956521739131
f1: 0.24005891016200295


In [19]:
confusion_matrix(y_val, classes)

array([[1509, 1930],
       [ 134,  326]])

*Модель выявила 314 клиентов с обратным откликом*

Выведите на экран в виде таблицы топ-6 признаков с наибольшими по модулю весами модели.

In [20]:
# your code here

weights = sorted(model.coef_[0], key=abs, reverse=True)
weights[:6]

[-0.2714424944474197,
 0.23747614833375216,
 -0.15010911812231525,
 -0.10328870470337817,
 -0.06170070569378883,
 -0.06170070569378883]

In [21]:
weights = weights[:6]

In [22]:
coef_table = pd.DataFrame({'features': list(X_train.columns), 'weights': list(model.coef_[0])})
coef_table.sort_values(by='weights')

features   weights
1  SOCSTATUS_WORK_FL -0.271442
5         DEPENDANTS -0.150109
3             GENDER -0.103289
7     LOAN_NUM_TOTAL -0.061701
8    LOAN_NUM_CLOSED -0.061701
2  SOCSTATUS_PENS_FL -0.050093
0                AGE -0.045314
6    PERSONAL_INCOME  0.000030
4        CHILD_TOTAL  0.237476

## Задание 2

Добавьте в Streamlit-приложение визуализацию результатов модели:

* опцию выбора порога и вывод метрик качества в зависимости от выбранного порога

* вывод прогноза модели на выбранном объекте (клиенте) - вероятность отклика на рекламу.

In [27]:
import pickle

In [28]:
with open('model.pickle', 'wb') as f:
  pickle.dump(model, f)

## Бонус

Попробуйте применить другие модели классификации для решения этой задачи (любые какие знаете).

Удалось ли добиться улучшения качества модели?

In [ ]:
# your code here

In [ ]:
# пришла рецензия на работу с стреамлитом. Поскольку курс разнообразен и уже давно забыл что нужно доделать 
#напоминаю себе: 1.Я не знаю как этот файл оказался в соседнем репозитории с чистым приложением (гит-баш наверное).

# 3. streamlit строит графики на ходу. Осталось узнать почему приложении выбрасывало ошибку при импотрте сиборна.
# 4. 
# 5. бд на postgres поднимай сразу в облаке из скриптов. из докера нифига не работает что на винде, что на деб, что в убунте. (не трать время.)
 #  и нет я сейчас в sqlite быстренько запросики посмотрю, то же не работает.

In [2]:
from sklearn import linear_model

In [3]:
df = pd.read_csv('datasets/bank_clients.csv')
# here we go again
X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']

In [4]:
 #посмотрим на наивного байеса

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
 
# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
 
# Инициализация Gaussian Naive Bayes classifier
gnb = GaussianNB()
 
# Tтренируем модель
gnb.fit(X_train, y_train)
 
# предсказания для теста
y_pred = gnb.predict(X_test)
 
# вычисление 
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')

Accuracy: 0.8636460782218423


In [8]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

print(f"Precision: {precision_score(y_test, y_pred)}")
print(f"Recall: {recall_score(y_test, y_pred)}")
print(f"f1: {f1_score(y_test, y_pred)}")

Precision: 0.23958333333333334
Recall: 0.08550185873605948
f1: 0.12602739726027398


In [15]:
df = pd.read_csv('datasets/bank_clients.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15594 entries, 0 to 15593
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   AGREEMENT_RK       15594 non-null  int64  
 1   AGE                15594 non-null  int64  
 2   SOCSTATUS_WORK_FL  15594 non-null  int64  
 3   SOCSTATUS_PENS_FL  15594 non-null  int64  
 4   GENDER             15594 non-null  int64  
 5   CHILD_TOTAL        15594 non-null  int64  
 6   DEPENDANTS         15594 non-null  int64  
 7   PERSONAL_INCOME    15594 non-null  float64
 8   LOAN_NUM_TOTAL     15594 non-null  int64  
 9   LOAN_NUM_CLOSED    15594 non-null  int64  
 10  TARGET             15594 non-null  int64  
dtypes: float64(1), int64(10)
memory usage: 1.3 MB


In [16]:
# we go again, and again
X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']
X_train, X_test, y_train, y_test = train_test_split(X , y, test_size=0.25, random_state=42)

In [17]:
from sklearn.tree import DecisionTreeClassifier

dtc = DecisionTreeClassifier()
dtc.fit(X_train, y_train)

pred_dtc = dtc.predict(X_test)


print(f'Accuracy: {accuracy_score(y_test, pred_dtc)}')


Accuracy: 0.8217491664529366


In [18]:
print(f"Precision: {precision_score(y_test, pred_dtc)}")
print(f"Recall: {recall_score(y_test, pred_dtc)}")
print(f"f1: {f1_score(y_test, pred_dtc)}")

Precision: 0.1671388101983003
Recall: 0.1282608695652174
f1: 0.14514145141451415


In [20]:
df = pd.read_csv('datasets/bank_clients.csv')
#df.info()
# again, and again
X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']
X_train, X_test, y_train, y_test = train_test_split(X , y, test_size=0.25, random_state=42)

In [21]:
# ну и конечно лес

In [22]:
from sklearn.ensemble import RandomForestClassifier

rfc = RandomForestClassifier()
rfc.fit(X_train, y_train)

pred_rfc = rfc.predict(X_test)

print(f'Accuracy: {accuracy_score(y_test, pred_rfc)}')

Accuracy: 0.8504744806360606


In [23]:
print(f"Precision: {precision_score(y_test, pred_rfc)}")
print(f"Recall: {recall_score(y_test, pred_rfc)}")
print(f"f1: {f1_score(y_test, pred_rfc)}")

Precision: 0.16756756756756758
Recall: 0.06739130434782609
f1: 0.09612403100775194


In [ ]:
# ....

In [ ]:
# не ну работа сдана, курс называется ЛИНЕЙНЫЕ Модели, занятие про интерпретацию еще впереди....

In [27]:
df = pd.read_csv('datasets/bank_clients.csv')
#df.info()
# again, and again
X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']
X_train, X_test, y_train, y_test = train_test_split(X , y, test_size=0.25, random_state=42)
X_train.shape, X_test.shape

((11695, 9), (3899, 9))

In [28]:
from sklearn.preprocessing import MinMaxScaler

mms = MinMaxScaler()
mms.fit(X_train) # вычислить min, max по каждому столбцу

X_train = pd.DataFrame(ss.transform(X_train), columns=X_train.columns)
X_test = pd.DataFrame(ss.transform(X_test), columns=X_test.columns)

X_train.head()

AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  GENDER  CHILD_TOTAL  \
0  0.152174                1.0                0.0     1.0          0.1   
1  0.782609                0.0                1.0     1.0          0.1   
2  0.239130                1.0                0.0     0.0          0.1   
3  0.891304                0.0                1.0     1.0          0.0   
4  0.282609                1.0                0.0     0.0          0.2   

   DEPENDANTS  PERSONAL_INCOME  LOAN_NUM_TOTAL  LOAN_NUM_CLOSED  
0    0.166667            0.060        0.090909         0.090909  
1    0.166667            0.018        0.272727         0.272727  
2    0.166667            0.080        0.090909         0.090909  
3    0.000000            0.012        0.090909         0.090909  
4    0.333333            0.038        0.181818         0.181818

In [ ]:
#Повторно построим модель логистической регрессии

In [29]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(max_iter=5000)

lr.fit(X_train, y_train)

pred = lr.predict(X_test)

In [30]:
importances = pd.DataFrame({'weights': lr.coef_[0], 'features': X_train.columns}).sort_values(by='weights')
importances

weights           features
0 -1.223560                AGE
7 -0.510138     LOAN_NUM_TOTAL
8 -0.510138    LOAN_NUM_CLOSED
2 -0.216640  SOCSTATUS_PENS_FL
3 -0.013605             GENDER
5  0.052033         DEPENDANTS
1  0.457631  SOCSTATUS_WORK_FL
4  1.064210        CHILD_TOTAL
6  5.023697    PERSONAL_INCOME

In [31]:
accuracy_score(y_test, pred)

0.8809951269556296

In [ ]:
## воот

In [ ]:
#import pickle

#with open('model_mms.mw', 'wb') as file:
#    pickle.dump(lr, file)